components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/cache/AuthenticationContextLoader.java

```diff
@@ -140,7 +141,10 @@ private void optimizeAuthenticatorConfig(AuthenticationContext context)
                 List<AuthenticatorConfig> authenticatorList = stepConfig.getAuthenticatorList();
                 for (AuthenticatorConfig authenticatorConfig : authenticatorList) {
                     authenticatorConfig.setIdPResourceIds(Collections.emptyList());
-                    authenticatorConfig.setApplicationAuthenticator(null);
+                    if (AuthenticatorPropertyConstants.DefinedByType.SYSTEM.equals(
+                            authenticatorConfig.getApplicationAuthenticator().getDefinedByType())) {
+                        authenticatorConfig.setApplicationAuthenticator(null);
+                    }
```

Comment:
why this system only check is needed ?

Reply:
This is happens at saving the caching. Previously, the authenticator set to null when saving the cache and resolve authenticator when retrieving from cache. For user defined authenticators, we need to load the authenticator config and then get the authenticator. So check whether authenticator is system defined, then set to null

Reply:
Suppose there's a user defined authenticator set in the context.
1. When adding to cache will it remove that instance ?
2. If (1) how does that get loaded
3. If not (1) it will serialize unnecessary object and grow at storage layer right ?